File: C++/103.cpp
#include <vector>
#include <iostream>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> result;
        vector<int> resultTmp;
        vector<TreeNode*> leftQueue;
        vector<TreeNode*> rightQueue;

        if (root == NULL)
            return result;
        leftQueue.push_back(root);
        if (root->left)
            rightQueue.push_back(root->left);
        if (root->right)
            rightQueue.push_back(root->right);
        while (!leftQueue.empty() || !rightQueue.empty()) {
            // 输出
            if (!leftQueue.empty()) {
                for (TreeNode *tmp: leftQueue) {
                    resultTmp.push_back(tmp->val);
                }
                result.push_back(resultTmp);
                resultTmp.clear();
            }
            if (!rightQueue.empty()) {
                for (int i = rightQueue.size()-1; i >= 0; i--) {
                    resultTmp.push_back(rightQueue.at(i)->val);
                }
                result.push_back(resultTmp);
                resultTmp.clear();
            }
            // 移动
            leftQueue.clear();
            if (!rightQueue.empty()) {
                for (TreeNode *tmp: rightQueue) {
                    if (tmp->left)
                        leftQueue.push_back(tmp->left);
                    if (tmp->right)
                        leftQueue.push_back(tmp->right);
                }
            }
            rightQueue.clear();
            if (!leftQueue.empty()) {
                for (TreeNode *tmp: leftQueue) {
                    if (tmp->left)
                        rightQueue.push_back(tmp->left);
                    if (tmp->right)
                        rightQueue.push_back(tmp->right);
                }
            }
        }
        return result;
    }
};

int main() {
    TreeNode a(3);
    TreeNode b(9);
    TreeNode c (20);
    TreeNode d(15);
    TreeNode e(7);
    a.left = &b;
    a.right = &c;
    c.left = &d;
    c.right = &e;
    Solution *s = new Solution();
    vector<vector<int>> result = s->zigzagLevelOrder(&a);
    for (int i = 0; i < result.size(); i++) {
        for (int j = 0; j < result.at(i).size(); j++) {
            cout << result.at(i).at(j) << " ";
        }
        cout << endl;
    }
}

File: Python/103.py
from typing import List

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        leftQueue = []; rightQueue = []; result = []; resultTmp = []
        if root is None:
            return result
        leftQueue.append(root)
        if root.left is not None:
            rightQueue.append(root.left)
        if root.right is not None:
            rightQueue.append(root.right)
        
        while len(leftQueue)>0 or len(rightQueue)>0:
            # 输出            
            for i in range(len(leftQueue)):
                resultTmp.append(leftQueue[i].val)
            if len(resultTmp) > 0:
                result.append(resultTmp)
            resultTmp = []
            for i in range(len(rightQueue)-1, -1, -1):
                resultTmp.append(rightQueue[i].val)
            if len(resultTmp) > 0:
                result.append(resultTmp)
            resultTmp = []
            # 移动
            leftQueue = []
            for i in range(len(rightQueue)):
                if rightQueue[i].left is not None:
                    leftQueue.append(rightQueue[i].left)
                if rightQueue[i].right is not None:
                    leftQueue.append(rightQueue[i].right)
            rightQueue = []
            for i in range(len(leftQueue)):
                if leftQueue[i].left is not None:
                    rightQueue.append(leftQueue[i].left)
                if leftQueue[i].right is not None:
                    rightQueue.append(leftQueue[i].right)
        return result

if __name__ == "__main__":
    a = TreeNode(3)
    b = TreeNode(9)
    c = TreeNode(20)
    d = TreeNode(15)
    e = TreeNode(7)
    a.left = b
    a.right = c
    c.left = d
    c.right = e
    s = Solution()
    result = s.zigzagLevelOrder(a)
    print(result)

        